framework/src/variables/MooseVariableFV.C

```diff
@@ -491,133 +505,320 @@ MooseVariableFV<OutputType>::getNeighborValue(const Elem * const neighbor,
-      };
+  // If we are performing a two term Taylor expansion for interpolary boundary faces (faces on
+  // boundaries that do not have associated Dirichlet conditions), then the element gradient depends
+  // on the boundary face value and the boundary face value depends on the element gradient, so we
+  // have a system of equations to solve. Here is the system:
+  //
+  // \nabla \phi_C - \frac{1}{V} \sum_{ibf} \phi_{ibf} \vec{S_f} =
+  //   \frac{1}{V} \sum_{of} \phi_{of} \vec{S_f}                       eqn. 1
+  //
+  // \phi_{ibf} - \vec{d_{Cf}} \cdot \nabla \phi_C = \phi_C            eqn. 2
+  //
+  // where $C$ refers to the cell centroid, $ibf$ refers to an interpolary boundary face, $of$
+  // refers to "other faces", e.g. non-ibf faces, and $f$ is a general face. $d_{Cf}$ is the vector
+  // drawn from the element centroid to the face centroid, and $\vec{S_f}$ is the surface vector,
+  // e.g. the face area times the outward facing normal
+
+  // We'll save off the interpolary boundary faces (ibf) for later assignment to the cache (these
+  // are the keys)
+  std::vector<const FaceInfo *> ibf_faces;
+  // ibf eqns: element gradient coefficients, e.g. eqn. 2, LHS term 2 coefficient
+  std::vector<VectorValue<Real>> ibf_grad_coeffs;
+  // ibf eqns: rhs b values. These will actually correspond to the elem_value so we can use a
+  // pointer and avoid copying. This is the RHS of eqn. 2
+  std::vector<const ADReal *> ibf_b;
+
+  // elem grad eqns: ibf coefficients, e.g. eqn. 1, LHS term 2 coefficients
+  std::vector<VectorValue<Real>> grad_ibf_coeffs;
+  // elem grad eqnas: rhs b value, e.g. eqn. 1 RHS
+  VectorValue<ADReal> grad_b = 0;
+
+  auto action_functor = [&volume_set,
+                         &volume,
+                         &elem_value,
+#ifndef NDEBUG
+                         &elem,
+#endif
+                         &ibf_faces,
+                         &ibf_grad_coeffs,
+                         &ibf_b,
+                         &grad_ibf_coeffs,
+                         &grad_b,
+                         this](const Elem & functor_elem,
+                               const Elem * const neighbor,
+                               const FaceInfo * const fi,
+                               const Point & surface_vector,
+                               Real coord,
+                               const bool elem_has_info) {
+    mooseAssert(fi, "We need a FaceInfo for this action_functor");
+    mooseAssert(elem == &functor_elem,
+                "Just a sanity check that the element being passed in is the one we passed out.");
+
+    if (isInterpolaryBoundaryFace(*fi))
+    {
+      if (_two_term_boundary_expansion)
+      {
+        ibf_faces.push_back(fi);
+
+        // eqn. 2
+        ibf_grad_coeffs.push_back(-1. * (elem_has_info
+                                             ? (fi->faceCentroid() - fi->elemCentroid())
+                                             : (fi->faceCentroid() - fi->neighborCentroid())));
```

Comment:
so for velocity on fully developped outlets, we should use the gradient computed on the boundary (15.141).
It is computed from the gradient at the centroid through 15.140. I dont think it'd be too bad to add here. 

We'd need to know when fi is at such a boundary though.

Reply:
Sigh. Yea this is rough. We have incompressible flow physics (modules) bleeding into generic computation of cell gradients/face values (framework). Going to have to think about how to do this intelligently.

Reply:
Yeah let's talk about it Wednesday again. There's #16585 to consider as well

Reply:
Well I'm probably going to be working on this today. I am planning to create an `INSFVVariable` class with #16585 in mind